File: metis-web/packages/web/server/sse.ts
import { Hono } from "hono";
import { getCookie } from "hono/cookie";
import { stream } from "hono/streaming";
import { config } from "./config.js";
import { logger } from "./logger.js";

export const sse = new Hono();

/**
 * SSE relay: GET /events -> metis-server GET /v1/events
 *
 * Dedicated handler for long-lived SSE connections that:
 * - Sets correct SSE headers (Content-Type, Cache-Control, Connection)
 * - Forwards Last-Event-ID from the browser to upstream for reconnection
 * - Injects the auth token from the HttpOnly cookie
 */
sse.get("/events", async (c) => {
  const token = getCookie(c, config.cookieName);
  if (!token) {
    logger.debug("sse: missing auth cookie");
    return c.json({ error: "not authenticated" }, 401);
  }

  const url = new URL(c.req.url);
  const targetUrl = `${config.metisServerUrl}/v1/events${url.search}`;

  const headers: Record<string, string> = {
    Authorization: `Bearer ${token}`,
    Accept: "text/event-stream",
  };

  // Forward Last-Event-ID for reconnection support
  const lastEventId =
    c.req.header("Last-Event-ID") ?? c.req.header("last-event-id");
  if (lastEventId) {
    headers["Last-Event-ID"] = lastEventId;
  }

  let upstreamResp: Response;
  try {
    upstreamResp = await fetch(targetUrl, { headers });
  } catch (err) {
    logger.error("sse: upstream fetch failed", {
      url: targetUrl,
      error: err instanceof Error ? err.message : String(err),
    });
    return c.json({ error: "upstream unreachable" }, 502);
  }

  if (!upstreamResp.ok) {
    logger.error("sse: upstream non-2xx response", {
      url: targetUrl,
      status: upstreamResp.status,
    });
    return c.json(
      { error: `upstream error: ${upstreamResp.status}` },
      upstreamResp.status as 500,
    );
  }

  if (!upstreamResp.body) {
    logger.error("sse: upstream returned no body", { url: targetUrl });
    return c.json({ error: "no upstream body" }, 502);
  }

  logger.debug("sse: connection established", { url: targetUrl });

  c.header("Content-Type", "text/event-stream");
  c.header("Cache-Control", "no-cache");
  c.header("Connection", "keep-alive");
  c.header("X-Accel-Buffering", "no");

  const reader = upstreamResp.body.getReader();

  return stream(c, async (s) => {
    try {
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        await s.write(value);
      }
    } catch {
      // Client disconnected or upstream closed — nothing to do
    } finally {
      reader.releaseLock();
    }
  });
});

File: metis-bff/src/sse.rs
use axum::{
    body::Body,
    extract::State,
    http::{header, StatusCode},
    response::{IntoResponse, Response},
};
use axum_extra::extract::cookie::CookieJar;
use http::Request;

use crate::state::BffState;
use crate::upstream::Upstream;

/// SSE relay handler for `/api/v1/events`.
///
/// Extracts the auth token (from auto_login_token or cookie), translates it
/// to a Bearer token, forwards to the upstream `/v1/events` endpoint, and
/// streams the SSE response back with proper headers.
pub async fn sse_relay<U: Upstream>(
    State(bff): State<BffState<U>>,
    jar: CookieJar,
    request: Request<Body>,
) -> impl IntoResponse {
    let token = match bff.resolve_token(&jar) {
        Some(t) => t,
        None => {
            return (
                StatusCode::UNAUTHORIZED,
                axum::Json(serde_json::json!({ "error": "not authenticated" })),
            )
                .into_response();
        }
    };

    let mut uri = "/v1/events".to_string();
    if let Some(query) = request.uri().query() {
        uri = format!("{uri}?{query}");
    }

    let mut builder = Request::builder()
        .method(request.method())
        .uri(&uri)
        .header(header::AUTHORIZATION, format!("Bearer {token}"));

    // Forward Last-Event-ID if present.
    if let Some(last_event_id) = request.headers().get("Last-Event-ID") {
        builder = builder.header("Last-Event-ID", last_event_id);
    }

    let upstream_req = match builder.body(Body::empty()) {
        Ok(req) => req,
        Err(_) => {
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                axum::Json(serde_json::json!({ "error": "failed to build request" })),
            )
                .into_response();
        }
    };

    match bff.upstream.forward(upstream_req).await {
        Ok(upstream_resp) => {
            // Stream the SSE response back with proper headers.
            let (parts, body) = upstream_resp.into_parts();
            let mut response = Response::from_parts(parts, body);

            // Ensure SSE content type and streaming headers.
            response
                .headers_mut()
                .insert(header::CONTENT_TYPE, "text/event-stream".parse().unwrap());
            response
                .headers_mut()
                .insert(header::CACHE_CONTROL, "no-cache".parse().unwrap());
            response
                .headers_mut()
                .insert(header::CONNECTION, "keep-alive".parse().unwrap());

            response.into_response()
        }
        Err(_) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            axum::Json(serde_json::json!({ "error": "internal error" })),
        )
            .into_response(),
    }
}
